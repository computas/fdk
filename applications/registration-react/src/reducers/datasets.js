import {
  DATASETS_REQUEST,
  DATASETS_SUCCESS,
  DATASETS_FAILURE
} from '../constants/ActionTypes';

export default function datasets(
  state = { isFetchingDatasets: false, datasetItems: null },
  action
) {
  switch (action.type) {
    case DATASETS_REQUEST:
      return {
        ...state,
        isFetchingDatasets: true
      };
    case DATASETS_SUCCESS:
      return {
        ...state,
        isFetchingDatasets: false,
        datasetItems: action.response.data
      };
    case DATASETS_FAILURE:
      return {
        ...state,
        isFetchingDatasets: false,
        datasetItems: null
      };
    default:
      return state;
  }
}
