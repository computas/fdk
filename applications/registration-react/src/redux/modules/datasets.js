import _ from 'lodash';
import {
  DATASETS_REQUEST,
  DATASETS_SUCCESS,
  DATASETS_FAILURE
} from '../../constants/ActionTypes';
import { fetchActions } from '../fetchActions';

function shouldFetch(metaState) {
  const threshold = 60 * 1000; // seconds
  return (
    !metaState ||
    (!metaState.isFetching &&
      (metaState.lastFetch || 0) < Date.now() - threshold)
  );
}

export function fetchDatasetsIfNeeded(catalogId) {
  return (dispatch, getState) =>
    shouldFetch(_.get(getState(), ['datasets', catalogId, 'meta'])) &&
    dispatch(
      fetchActions(`/catalogs/${catalogId}/datasets?size=100`, [
        { type: DATASETS_REQUEST, meta: { catalogId } },
        { type: DATASETS_SUCCESS, meta: { catalogId } },
        { type: DATASETS_FAILURE, meta: { catalogId } }
      ])
    );
}

const initialState = {};

export default function datasets(state = initialState, action) {
  switch (action.type) {
    case DATASETS_REQUEST:
      return {
        ...state,
        [action.meta.catalogId]: {
          meta: {
            isFetching: true,
            lastFetch: null
          }
        }
      };
    case DATASETS_SUCCESS:
      return {
        ...state,
        [action.meta.catalogId]: {
          items: _.get(action.payload, ['_embedded', 'datasets']),
          meta: {
            isFetching: false,
            lastFetch: Date.now()
          }
        }
      };
    case DATASETS_FAILURE:
      return {
        ...state,
        [action.meta.catalogId]: {
          meta: {
            isFetching: false,
            lastFetch: null
          }
        }
      };
    default:
      return state;
  }
}

export const getDatasetItemsByCatalogId = (datasets, catalogId) =>
  _.get(datasets, [catalogId, 'items']);
