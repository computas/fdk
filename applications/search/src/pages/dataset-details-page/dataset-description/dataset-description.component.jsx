import React from 'react';
import PropTypes from 'prop-types';

import localization from '../../../lib/localization';
import { ShowMore } from '../../../components/show-more/show-more';
import { PublisherLabel } from '../../../components/publisher-label/publisher-label.component';
import { HarvestDate } from '../../../components/harvest-date/harvest-date.component';
import { SearchHitHeader } from '../../../components/search-hit-header/search-hit-header.component';

export class DatasetDescription extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      showAll: false
    };
    this.toggleShowAll = this.toggleShowAll.bind(this);
  }

  toggleShowAll() {
    this.setState({ showAll: !this.state.showAll });
  }

  _renderPublisher() {
    const { publisher } = this.props;

    if (!publisher) {
      return null;
    }
    return (
      <PublisherLabel
        label={localization.search_hit.owned}
        publisherItem={publisher}
      />
    );
  }

  render() {
    const { harvest, title, publisher, themes, provenance } = this.props;
    return (
      <header>
        <div className="fdk-detail-date mb-5">
          <HarvestDate harvest={harvest} />
        </div>

        <SearchHitHeader
          title={title}
          publisherLabel={localization.search_hit.owned}
          publisher={publisher}
          theme={themes}
          provenance={provenance}
        />

        {this.props.description && (
          <ShowMore
            showMoreButtonText={localization.showFullDescription}
            label={localization.description}
            contentHtml={this.props.descriptionFormatted}
          />
        )}

        {this.props.objective && (
          <ShowMore
            showMoreButtonText={localization.showFullObjective}
            label={localization.objective}
            contentHtml={this.props.objective}
          />
        )}
      </header>
    );
  }
}

DatasetDescription.defaultProps = {
  title: '',
  description: '',
  descriptionFormatted: null,
  objective: '',
  publisher: null,
  themes: null,
  harvest: null,
  provenance: null
};

DatasetDescription.propTypes = {
  title: PropTypes.string,
  description: PropTypes.string,
  descriptionFormatted: PropTypes.string,
  objective: PropTypes.string,
  publisher: PropTypes.object,
  themes: PropTypes.array,
  harvest: PropTypes.object,
  provenance: PropTypes.object
};
