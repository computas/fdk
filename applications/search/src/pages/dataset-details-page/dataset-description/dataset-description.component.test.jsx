import React from 'react';
import { shallow } from 'enzyme';
import { DatasetDescription } from './dataset-description.component';
import datasets from '../../../../test/fixtures/datasets';

test('should render DatasetDescription correctly with no props', () => {
  const minWrapper = shallow(<DatasetDescription />);
  expect(minWrapper).toMatchSnapshot();
});

test('should render DatasetDescription correctly', () => {
  const { _source } = datasets[0];
  const defaultProps = {
    title: _source.title.nb,
    description: _source.description.nb,
    objective: _source.objective.nb,
    publisher: _source.publisher,
    themes: _source.theme,
    harvest: _source.harvest
  };
  const wrapper = shallow(<DatasetDescription {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render DatasetDescription correctly', () => {
  const { _source } = datasets[1];
  const defaultProps = {
    title: _source.title.nb,
    description: _source.description.nb,
    objective: _source.objective.nb,
    publisher: _source.publisher,
    themes: _source.theme,
    harvest: _source.harvest
  };
  const wrapper = shallow(<DatasetDescription {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});
