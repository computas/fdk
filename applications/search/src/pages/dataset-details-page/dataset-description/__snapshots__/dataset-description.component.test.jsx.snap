// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render DatasetDescription correctly 1`] = `
<header>
  <div
    className="fdk-detail-date mb-5"
  >
    <HarvestDate
      harvest={null}
    />
  </div>
  <SearchHitHeader
    provenance={null}
    publisher={
      Object {
        "id": "987654321",
        "name": "TESTETATEN",
        "uri": "http://data.brreg.no/enhetsregisteret/enhet/987654321",
      }
    }
    publisherItems={null}
    publisherLabel="Owned by"
    tag="h1"
    theme={
      Array [
        Object {
          "title": Object {
            "nb": "Forvaltning og offentlig støtte",
          },
          "uri": "http://publications.europa.eu/resource/authority/data-theme/GOVE",
        },
        Object {
          "title": Object {
            "nb": "Miljø",
          },
          "uri": "http://publications.europa.eu/resource/authority/data-theme/ENVI",
        },
      ]
    }
    title="AAMarkagrensen Oslo Kommune og nærliggende kommuner"
    titleLink={null}
  />
  <ShowMore
    contentHtml={null}
    label="Description"
    showLessButtonText="Show less"
    showMoreButtonText="Show the full description"
  />
  <ShowMore
    contentHtml="Datasettes formål er nullam quis rius eget urna mollis ornare vel eu leo. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Vestibulum id ligula porta felis euismod semper."
    label="Objective"
    showLessButtonText="Show less"
    showMoreButtonText="Show the full objective"
  />
</header>
`;

exports[`should render DatasetDescription correctly 2`] = `
<header>
  <div
    className="fdk-detail-date mb-5"
  >
    <HarvestDate
      harvest={null}
    />
  </div>
  <SearchHitHeader
    provenance={null}
    publisher={
      Object {
        "id": "987654321",
        "name": "TESTETATEN",
        "uri": "http://data.brreg.no/enhetsregisteret/enhet/987654321",
      }
    }
    publisherItems={null}
    publisherLabel="Owned by"
    tag="h1"
    theme={
      Array [
        Object {
          "title": Object {
            "nb": "Forvaltning og offentlig støtte",
          },
          "uri": "http://publications.europa.eu/resource/authority/data-theme/GOVE",
        },
        Object {
          "title": Object {
            "nb": "Miljø",
          },
          "uri": "http://publications.europa.eu/resource/authority/data-theme/ENVI",
        },
      ]
    }
    title="AAMarkagrensen Oslo Kommune og nærliggende kommuner"
    titleLink={null}
  />
  <ShowMore
    contentHtml={null}
    label="Description"
    showLessButtonText="Show less"
    showMoreButtonText="Show the full description"
  />
  <ShowMore
    contentHtml="Datasettes formål er nullam quis rius eget urna mollis ornare vel eu leo. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras mattis consectetur purus sit amet fermentum. Vestibulum id ligula porta felis euismod semper."
    label="Objective"
    showLessButtonText="Show less"
    showMoreButtonText="Show the full objective"
  />
</header>
`;

exports[`should render DatasetDescription correctly with no props 1`] = `
<header>
  <div
    className="fdk-detail-date mb-5"
  >
    <HarvestDate
      harvest={null}
    />
  </div>
  <SearchHitHeader
    provenance={null}
    publisher={null}
    publisherItems={null}
    publisherLabel="Owned by"
    tag="h1"
    theme={null}
    title=""
    titleLink={null}
  />
</header>
`;
